pallets/creditcoin/src/tests/cleanup.rs

```diff
@@ -0,0 +1,289 @@
+					<$storage>::storage_double_map_final_key1(first)
+				}
+			}
+		)*
+	};
+}
+
+trait GetPartialKey {
+	fn partial_key(exp: u64) -> Vec<u8>;
+}
+
+impl_get_partial_key!(AskOrders, BidOrders, Offers);
+
+fn cursor_for<Storage: GetPartialKey>(
+	first: u64,
+) -> Option<WeakBoundedVec<u8, ConstU32<{ StorageItemCleanupState::<BlockNumber>::MAX_CURSOR_LEN }>>>
+{
+	Some(weak_bounded(Storage::partial_key(first)))
+}
+
+#[test]
+fn cleans_up_asks_to_limit() {
+	let count = LIMIT * 2;
+	let exp = 5;
+
+	StorageBuilder::new()
+		.ask_orders(count, exp)
+		.finish()
+		.then_run(|| {
+			assert_eq!(AskOrders::iter().count(), count);
+
+			cleanup(exp);
+
+			assert_eq!(AskOrders::iter().count(), count - LIMIT);
+			assert_eq!(
+				cleanup_state(),
+				Some(StorageCleanupState {
+					ask_orders: StorageItemCleanupState {
+						on_block: exp,
+						cursor: cursor_for::<AskOrders>(exp)
+					},
+					..StorageCleanupState::new(exp)
+				})
+			);
+		})
+		.then_run(|| {
+			assert_eq!(AskOrders::iter().count(), count - LIMIT);
+
+			cleanup(exp + 1);
+
+			assert_eq!(AskOrders::iter().count(), 0);
+			assert_eq!(
+				cleanup_state(),
+				Some(StorageCleanupState {
+					ask_orders: StorageItemCleanupState::new(exp + 1),
+					..StorageCleanupState::new(exp)
+				})
+			);
+		})
+		.then_run(|| {
+			assert_eq!(AskOrders::iter().count(), 0);
+
+			cleanup(exp + 2);
+
+			assert_eq!(AskOrders::iter().count(), 0);
+			assert_eq!(cleanup_state(), Some(StorageCleanupState::new(exp + 3)));
+		});
+}
+
+#[test]
+fn cleans_up_bids() {
+	let count = LIMIT / 2;
+	let exp = 5;
+
+	StorageBuilder::new().bid_orders(count, exp).finish().then_run(|| {
+		assert_eq!(BidOrders::iter().count(), count);
+
+		cleanup(exp);
+
+		assert_eq!(BidOrders::iter().count(), 0);
+		assert_eq!(cleanup_state(), Some(StorageCleanupState::new(exp + 1)));
+	});
+}
+
+#[test]
+fn cleans_up_offers() {
+	let count = LIMIT / 2;
+	let exp = 5;
+
+	StorageBuilder::new().offers(count, exp).finish().then_run(|| {
+		assert_eq!(Offers::iter().count(), count);
+
+		cleanup(exp);
+
+		assert_eq!(Offers::iter().count(), 0);
+		assert_eq!(cleanup_state(), Some(StorageCleanupState::new(exp + 1)));
+	});
+}
+
+#[test]
+fn cleans_up_to_limit() {
+	let count = LIMIT / 2;
+	let exp = 5;
+
+	StorageBuilder::new()
+		.ask_orders(count, exp)
+		.bid_orders(count, exp)
+		.offers(count, exp)
+		.finish()
+		.then_run(|| {
+			assert_eq!(AskOrders::iter().count(), count);
+			assert_eq!(BidOrders::iter().count(), count);
+			assert_eq!(Offers::iter().count(), count);
+
+			cleanup(exp);
+
+			assert_eq!(AskOrders::iter().count(), 0);
+			assert_eq!(BidOrders::iter().count(), 0);
+			assert_eq!(Offers::iter().count(), count);
+
+			assert_eq!(
+				cleanup_state(),
+				Some(StorageCleanupState {
+					ask_orders: StorageItemCleanupState::new(exp + 1),
+					bid_orders: StorageItemCleanupState::new(exp + 1),
+					offers: StorageItemCleanupState { on_block: exp, cursor: None },
+				})
+			);
+		});
+}
```

Comment:
Out of curiosity: how many blocks does it take to remove 1M * Ask, Bid, Offer ?

Reply:
The limit is about 43000 entries per block, so `3000000/43000=70` blocks or so. The current state of mainnet takes about 40 blocks to remove Aella's currently recorded entries.